File: Data Structures & Algorithms/two-integer-sum/submission-6.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        visited = {}

        for index, num in enumerate(nums):
            result = target - num

            if result in visited:
                return [visited[result], index]
            else:
                visited[num] = index

        return [-1, -1]

File: Data Structures & Algorithms/combination-target-sum/submission-6.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> result = new ArrayList<>();
        dfs(nums, target, 0, new ArrayList<>(), result);
        return result;
    }

    public void dfs(int[] nums, int target, int i, List<Integer> current, List<List<Integer>> result){

        if(target == 0){
            result.add(new ArrayList<>(current));
            return;
        }

        if(target < 0 || i >= nums.length){
            return;
        }

        current.add(nums[i]);
        dfs(nums, target - nums[i], i, current, result);
        current.remove(current.size() - 1);
        dfs(nums, target, i + 1, current, result);
    }
}
